src/v/archival/ntp_archiver_service.cc

```diff
@@ -2613,6 +2664,72 @@ ss::future<> ntp_archiver::apply_spillover() {
     }
 }
 
+flush_result ntp_archiver::flush() {
+    // Return early if we are not the leader.
+    if (!_parent.is_leader()) {
+        vlog(_rtclog.debug, "in flush(), not leader.");
+        return flush_result::not_leader;
+    }
+
+    _flush_uploads_offset = model::prev_offset(
+      max_uploadable_offset_exclusive());
+    vlog(
+      _rtclog.debug,
+      "accepted flush(), flush offset is {}.",
+      _flush_uploads_offset.value());
+    return flush_result::accepted;
+}
+
+ss::future<wait_result> ntp_archiver::wait() {
+    // If there hasn't been a flush request, return early.
+    if (!flush_in_progress()) {
+        vlog(_rtclog.debug, "In wait(), flush() is not in progress.");
+        co_return wait_result::not_in_progress;
+    }
+
+    // If our manifest's insync offset is greater than the user provided offset,
+    // we have uploaded or flush'ed() up to this point.
+    if (current_flush_complete()) {
+        vlog(_rtclog.debug, "In wait(), flush() is already complete.");
+        co_return wait_result::complete;
+    }
+
+    // If this partition has lost leadership, we indicate so, and the user
+    // may want to reissue a flush() command to the partition leader.
+    if (lost_leadership()) {
+        vlog(_rtclog.debug, "In wait(), leadership was already lost.");
+        co_return wait_result::lost_leadership;
+    }
+
+    // Otherwise, we will wait on the condition variable to signal.
+    try {
+        vlog(_rtclog.debug, "Waiting for flush() to complete");
+        co_await _flush_cond.wait();
+    } catch (const ss::broken_condition_variable&) {
+        // We shutdown while waiting for a flush() to finish.
+        // Return a failure result.
+        co_return wait_result::failed;
+    }
+
+    // Recheck flush status.
+    if (current_flush_complete()) {
+        vlog(_rtclog.debug, "After waiting in wait(), flush was completed.");
+        _flush_uploads_offset.reset();
```

Comment:
It's a little surprising for a wait method to update state. Did you consider having callers pass in an offset to wait on, and having the upload loop reset the flushed offset? It may be more intuitive for a caller, as it'd be a bit clearer what they're waiting for

Reply:
Updated function signature and behaviour in `wait()`.